src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs

```diff
@@ -113,118 +113,81 @@ public async Task<Solution> CreateSolutionAsync(Checksum newSolutionChecksum, Ca
+                    if (!newProjectIds.Object.Contains(oldProjectId))
+                        solution = solution.RemoveProject(oldProjectId);
                 }
 
-                return solution;
+                // Fourth, update any changed projects.
 
-                async Task PopulateOldProjectMapAsync()
+                // We do this in two stages.  The initial stage removes all project references from projects that
+                // changed. this ensures exceptions will not occur for cyclic references during an incremental update.
+                foreach (var (projectId, oldProjectStateChecksums, newProjectStateChecksums) in changedProjects)
                 {
-                    foreach (var (projectId, projectState) in solution.SolutionState.ProjectStates)
-                    {
-                        var projectChecksums = await projectState.GetStateChecksumsAsync(cancellationToken).ConfigureAwait(false);
-                        if (olds.Object.Contains(projectChecksums.Checksum))
-                            oldMap.Add(projectId, projectChecksums);
-                    }
+                    if (oldProjectStateChecksums.ProjectReferences.Checksum != newProjectStateChecksums.ProjectReferences.Checksum)
+                        solution = solution.WithProjectReferences(projectId, SpecializedCollections.EmptyEnumerable<ProjectReference>());
                 }
 
-                async Task PopulateNewProjectMapAsync(SolutionCreator @this)
+                // The final stage actually goes through and update the projects.
+                foreach (var (projectId, oldProjectStateChecksums, newProjectStateChecksums) in changedProjects)
                 {
-                    var projectStateChecksums = await @this._assetProvider.GetAssetsAsync<ProjectStateChecksums>(
-                        assetHint: AssetHint.None, news.Object, cancellationToken).ConfigureAwait(false);
-
-                    foreach (var (_, projectStateChecksum) in projectStateChecksums)
-                        newMap.Add(projectStateChecksum.ProjectId, projectStateChecksum);
+                    solution = await UpdateProjectAsync(
+                        solution.GetRequiredProject(projectId), oldProjectStateChecksums, newProjectStateChecksums, cancellationToken).ConfigureAwait(false);
                 }
-            }
 
-            private async ValueTask SynchronizeAssetsAsync(Dictionary<ProjectId, ProjectStateChecksums> oldMap, Dictionary<ProjectId, ProjectStateChecksums> newMap, CancellationToken cancellationToken)
-            {
-                // added project
-                foreach (var (projectId, projectStateChecksums) in newMap)
-                {
-                    if (oldMap.ContainsKey(projectId))
-                        continue;
-
-                    await _assetProvider.SynchronizeProjectAssetsAsync(projectStateChecksums, cancellationToken).ConfigureAwait(false);
```

Comment:
note: the call to synchronize is unnecessary.  CreateProjectInfoAsync already does that.